### ISSUE TITLE
breakpoints

### ISSUE THREAD

**Issue by krakenlake** [krakenlake/vmon#12]

**Comment by benmesander** [comment]
I was thinking about this - perhaps the first steps in this direction would be to save the RV32I registers (what the monitor currently uses, unless we port to RV32E) upon entry.  And in the 'g' command, restore the registers before jumping to the address. And add a command in the monitor to print the register contents. And show the entrypoint to return back to the monitor maybe in the 'i' command. 

This would allow users to initially set breakpoints by replacing an instruction with a jump to the monitor. They would have to save the instruction that was overwritten and restore it when they wanted to remove the breakpoint. The next step would be to automate this with a command. 

Thoughts?

**Comment by krakenlake** [comment]
Yes, that could work. But maybe we have to work with exceptions as well - idk. There's probably already a documented preferred way to do this in RISC-V, I just haven't seen that yet.

**Comment by benmesander** [comment]
I wondered the same. It might not be available on all real world hardware though. I also was wondering about single-stepping. Normally monitors offer that feature (usually two flavors, one that steps over calls, one that steps into calls). I did all this on a Z-80 back in the day, so surely it's got to be easier on a processor like RISC-V.

**Comment by krakenlake** [comment]
I've researched this a bit now, I guess temporarily replacing instructions with `ebreak` is the way to go. 
https://www2.lauterbach.com/pdf/debugger_riscv.pdf

**Comment by benmesander** [comment]
Will have to read this. Seems like a hassle for compressed instructions

**Comment by benmesander** [comment]
yeah. there's a c.ebreak insn

**Comment by krakenlake** [comment]
yes, see `insn_is_compressed`  in `decode.S`:
